game-app/game-core/src/main/java/games/strategy/triplea/ui/AbstractUndoableMovesPanel.java

```diff
@@ -137,13 +137,27 @@ private JComponent newComponentForMove(final AbstractUndoableMove move) {
     final Box textBox = new Box(BoxLayout.X_AXIS);
     textBox.add(text);
     textBox.add(Box.createHorizontalGlue());
-    final JButton cancelButton = new JButton(new UndoMoveActionListener(move.getIndex()));
-    setSize(buttonSize, cancelButton);
+    final JButton undoButton = new JButton("Undo");
+    final int moveIndex = move.getIndex();
+    undoButton.addActionListener(
+        (e) -> {
+          // Record position of scroll bar as percentage.
+          scrollBarPreviousValue = scroll.getVerticalScrollBar().getValue();
+          final String error = movePanel.undoMove(moveIndex);
+          if (error == null) {
+            // Disable the button so it can't be clicked again until the UI is updated.
+            undoButton.setEnabled(false);
```

Comment:
Just so I understand it correctly:
`movePanel.undoMove` causes the button to disappear somehow (presumably because the backing panel disappears), so when the same panel is shown again the button is newly created so it can be clicked again.
Is this assumption correct?

Reply:
There's a button for each move. So this button will disappear when the move is no longer shown in the list. But that's not synchronous so we need to make the button not clickable in the meantime so it's not pressed a second time.

The button doesn't get recreated because the move doesn't exist. But any other moves still have their own buttons and any new moves will get them too.

Reply:
Okay, but which part does manipulate the list?

Reply:
Basically I'm not sure why the code does what you say it does. For me this looks like the button is disabled and for some reason there's no problem for it to be disabled forever.

So maybe this is just me, but if my confusion is justified I'd appreciate an explaining comment there, or an adjustment to the existing comment

Reply:
Sure, let me expand the comment. But TL;DR is the whole content will be re-created via setMoves() which will be called after an update.

Reply:
Improved the comment.